### ISSUE TITLE
Permissions Error Android

### ISSUE THREAD

**Issue by LeviWilliams** [mrousavy/react-native-google-nearby-messages#11]
Currently have been trying to get this library working on Android with no luck. First thing I discovered is that if google play services is out of date the addOnErrorListener will return a permissions error and the app will crash. Even if the app has full permissions. Often (but not every time) on first load the app will prompt the user to give the app correct permissions if they have not already been granted, then the same error will persist again.

Also, if the app is downloaded and given full permissions before first load then the app will send subscribe/publish once. Then on reload of the app it will no longer work and persist the same error as the first time.

I have followed the troubleshooting guide for Android. The api key is correct (and working because it works on first load) and permissions are given yet I cannot get it to work on Android. Have you had any similar issues?

Device: Huawei P20 Lite running Android 9.

**Comment by LeviWilliams** [comment]
Also will test on a Samsung Galaxy Tab E later and see if the error persists.

**Comment by LeviWilliams** [comment]
Forgot to mention iOS Devices work fine. Will also test this issue on a Samsung Galaxy Tab E later to see if the issue persists.

**Comment by mrousavy** [comment]
Okay that's really weird, unfortunately I don't have an android device right now, I'll try to get one from somewhere asap.

The permission flow is really hard to keep track off, could you in the meantime maybe try starting the app in debug mode in android studio and reproduce the problem? Knowing which line of code is the problem will help me fix this problem.

I'm guessing the pre-condition checks in the connect() function on android are not perfect, I'll look into those. The official docs over at https://developers.google.com/nearby/messages/overview don't really go into much detail considering edge cases such as permissions, missing BLE hardware and special devices..

If you're experienced with android, I'd glady accept any help! :)

**Comment by LeviWilliams** [comment]
Yes I've been looking into it, at first I figured it must just be an outdated play services problem but its seems to be more than that. 

I've done everything from reinstall the app, to update the phone, to remove and reset play services. It doesn't seem like its actually a permissions error but I could be wrong. I will try to reproduce in Android Studio.

To clarify as well - when I console log "kind, hasError, message" from the error listener I get: "PERMISSIONS ERROR false undefined" and doing an additional check from react native permissions returns that I do have permissions.

**Comment by LeviWilliams** [comment]
On the Android emulator it also doesn't get passed the checking of permissions phase, as expected though right? I don't think it can run on there regardless.

On the Samsung tablet so far I can't even get the app to load, it's crashing due to this lib. Gonna do some more digging there and see if it's related.

**Comment by LeviWilliams** [comment]
In regards to the tablet, after lots of updates and reinstalling/clearing cache I got it to load up but I get the same error cited in the orginal post. Seems to be consistent between devices and Android versions.

**Comment by JoaoPV** [comment]
Hi, just created a project with "react-native": "0.62.2" and using  'com.google.android.gms:play-services-nearby:17.0.0',  but i'm getting the same error as above:  console.error : 'PERMISSIONS ERROR false' when trying to connect.

**Comment by mrousavy** [comment]
So the PERMISSION_ERROR comes from here:

https://github.com/mrousavy/react-native-google-nearby-messages/blob/master/android/src/main/java/com/mrousavy/nearby/GoogleNearbyMessagesModule.kt#L73

Which is stupid, since it doesn't have an error and should work anyways. I should definetly get rid of the `hasError` property, if it doesn't have an error just don't emit the event.

1. does your android device pass these lines of code in the connect function?:

https://github.com/mrousavy/react-native-google-nearby-messages/blob/master/android/src/main/java/com/mrousavy/nearby/GoogleNearbyMessagesModule.kt#L47-L58

> (If you can connect without a promise rejection/error in javascript you get past these lines of code)

2. Does the `publish` function reach this line of code:

https://github.com/mrousavy/react-native-google-nearby-messages/blob/master/android/src/main/java/com/mrousavy/nearby/GoogleNearbyMessagesModule.kt#L171

> (If you can publish without a promise rejection/error in javascript you can reach this line)

Thanks!

**Comment by mrousavy** [comment]
@LeviWilliams @JoaoPV I've just pushed https://github.com/mrousavy/react-native-google-nearby-messages/commit/b1693034347a5b51c016968e185a042f14afe712 - would be awesome if you could try running the `example/` app on your android device.

In the meantime, I'll try to get my hands on an android device.

**Comment by LeviWilliams** [comment]
I was unable to get the test repo running on my android device - all kinds of random errors unrelated to this issue. Wasted way too much time trying to get it working to no avail. 

I don't think this is a big breaking change anyway shouldn't we be good to push it up npm/yarn?

**Comment by mrousavy** [comment]
@LeviWilliams I've published it to npm, please check it out if that works for me. Be aware that I have not tested it!

**Comment by JoaoPV** [comment]
@LeviWilliams  i will try today.

**Comment by LeviWilliams** [comment]
Tested the new version along with the hooks on both android devices, works fine!

**Comment by LeviWilliams** [comment]
This removed the error cited initially. Now I am getting "The Messages Client was null. Did GoogleNearbyMessagesModule native constructor fail to execute?" along with crashes on Android. Trying to pinpoint the error and if this is something completely different.

**Comment by mrousavy** [comment]
@LeviWilliams The connect() function initializes the messages client, so there has to be an error being thrown before the client could be initialized. Maybe the error message was right after all? I'll look into this tomorrow, it's 1am in austria rn

**Comment by LeviWilliams** [comment]
@mrousavy Sounds good, can't quite pinpoint what this issue specifically is. I got that to log twice but most of the time the app just crashes and the logs don't indicate much.

**Comment by mrousavy** [comment]
Yep, it's an error thrown in `connect()`. Are you sure you called `connect()` before calling `publish()` or `subscribe()`? (My hooks will do this for you)
Because the message client can only be null when either `connect` was not called, or there was an error thrown in `connect` (put a try catch around).

I'm looking into this. I think I'll get my hands on an android device next week.

**Comment by LeviWilliams** [comment]
Yes, the error only persists on Android. Using the hooks produces no issues on iOS. I am using the hooks in this order: publication, error callback, and isNearby. Also had intermittent crashes on the android emulators. I'll do some more isolated debugging and see what I can find.

**Comment by LeviWilliams** [comment]
Hey everyone, with more testing done I've narrowed it down to the publish function. Went error hunting in the hooks and also just using the functions themselves.

Connect and subscribe seem fine but the app freezes or crashes once it hits the publish function. I was unable to find any specific type of error message to indicate what is actually happening. Maybe the disconnect isn't actually unpublishing? Thoughts? @mrousavy

**Comment by mrousavy** [comment]
@LeviWilliams 
The `publish` function does not contain enough logic on my side that it can cause a crash. 

The thing is, I'm calling the publish function from the official Google Nearby Messages lib, and add an `onCompleteListener`. Apparently, Google's func doesn't complete at all, that's why you're seeing freezes. I reject and resolve the promise at any point the function could possibly return in. See: [::publish](https://github.com/mrousavy/react-native-google-nearby-messages/blob/master/android/src/main/java/com/mrousavy/nearby/GoogleNearbyMessagesModule.kt#L160-L182)

What exactly do you mean by 
> Maybe the disconnect isn't actually unpublishing? 

Does the first `publish()` work, but the second doesn't anymore? If that's the case, this is a totally different problem.

1. Does it crash instantly once you call publish?
2. How long are the freezes approximately?
3. Can you debug **your** app in Android Studio? There are more useful error informations there, and you could also add breakpoints in my code, since it's just raw Java (Kotlin) code in the `node_modules`, which would help me a lot.

**Comment by LeviWilliams** [comment]
@mrousavy 

Before I dug around too deep in Android Studio and switched over to version 1.0.12. Now I'm getting the exact same type of freezing in the subscribe hook instead of publish. I'm guessing the real problem was always subscribe but for some reason my initial debugging indicated otherwise. The freeze is around 30sec-1 min sec on average and only takes place when another device is actually in the area.

Here is the flow: open app first device -> app is fine -> open app on another device (ios or android) -> app immediately freezes.

I debugged the hook and the last thing I'm able to log before it freezes is that it is connected and subscribed. An iOS device is able to receive the Android device's message before it freezes out leading me to believe there is a problem in how Android is handling receiving a message? Everything still works as expected between iOS devices.

**Comment by LeviWilliams** [comment]
Here is an example of the flow of me opening an ios device first and then the android device, the message is lost on ios because the app crash on android.

![image](https://user-images.githubusercontent.com/32551245/85617618-cba2b900-b61c-11ea-8448-c60dd0192b6f.png)

**Comment by mrousavy** [comment]
![image](https://user-images.githubusercontent.com/15199031/85837902-029dd980-b799-11ea-9326-daa820c68096.png)

Maybe it's because of this line? Can you try changing this to `.BLUETOOTH`, `.BLE`, `.DEFAULT`, `.NONE` and see if that makes any difference? (It's in the `connect` function in `GoogleNearbyMessagesModule.kt`)

**Comment by mrousavy** [comment]
@LeviWilliams I've just published 1.0.13 on npm. It requires a new API for the connect function, which allows you to change those modes in the comment above. Try tweaking those options and use different `discoveryMediums`, then tell me which one has worked. Unfortunately I couldn't find any docs on the Google Nearby Messages API page which explain how to change the discovery mediums on android, that's why I'm using this `zze()` method (which I'm 100% sure doesn't do what I expect it to do).

With the correct input we can map the values to their correct integer, e.g.: `ble` -> `2`, `audio` -> `3`, ...

I've dug around the source code, which looked like `audio` or `microphone` should be of value `4` as a discovery medium, while `ble` is `2`. Still don't know where to set that.

**Comment by mrousavy** [comment]
I've rewritten the Strategy Builder from google since they somehow forgot to add a setter for the discovery medium. What are they doing over there?

Anyways, v1.0.14 is on npm, and I've tested it with a friend - we fixed the crash but the `subscribing()` func never completes. Maybe I'm adding the wrong `onCompleteListener`..? The error messages tells something about a Google Play Services error, which is BS as far as I can tell.

You can try running:

```ts
//                        Supported values on android are:
//                          ['ble', 'bluetooth', 'none', 'default', 'audio']
connect({ discoveryMediums: ['ble', 'bluetooth'] });
// then subscribe and/or publish
```

and see if that works (typescript will tell you that only 'ble' and 'audio' are valid, but it works. it's just private for now.)

**Comment by LeviWilliams** [comment]
@mrousavy v1.0.14 broke my iOS app even, "excessive number of pending callbacks" causing app deadlock and crashes. I'm just using the same hooks as before. Also can't get the app running on Android.

**Comment by mrousavy** [comment]
@LeviWilliams can you share your code so I can test your example? I got it working on iOS without any problems. (v1.0.18)

I'm even using the hooks. Just be aware that by passing in the `NearbyConfig` you have to memoize it using `useMemo` so it does not get re-created every re-render and therefore trigger infinite re-renders.

See: [the hooks docs in the README](https://github.com/mrousavy/react-native-google-nearby-messages#react-hooks)

**Comment by mrousavy** [comment]
@LeviWilliams can you test again with v1.0.20? Accidentally referenced the wrong `onFound` in my callback causing the StackOverflow exception, which is fixed now thanks to @wimvdb (see: #17)

**Comment by LeviWilliams** [comment]
@mrousavy all bugs mentioned in this thread fixed in version 1.0.20 - closing this issue.

**Comment by mrousavy** [comment]
finally lmao

thanks for your help @LeviWilliams
